### PR TITLE
Add exit this page component

### PR DESCRIPTION
**Pull request by owenatgov**
## What/Why
Adds the Exit This Page component to GOV.UK Frontend. Details about this component including functionality, history of work and decisions can be found in the epic: https://github.com/alphagov/govuk-design-system/issues/1755

### Standard view
Large screen:
Screenshot 2023-03-28 at 13 59 20

Small screen:
Screenshot 2023-03-28 at 13 59 44

### With indicator light active
Large screen:
Screenshot 2023-03-28 at 14 00 02

Small screen:
Screenshot 2023-03-28 at 14 00 40

### Forced colours mode (Windows High Contrast)
Large screen:
Screenshot 2023-03-28 at 14 02 58

Small screen:
Screenshot 2023-03-28 at 14 03 36

### Forced colours with indicator light active
Large screen:
Screenshot 2023-03-28 at 14 05 04

Small screen:
Screenshot 2023-03-28 at 14 04 29

## Testing
This component has been tested extensively via the following means:

- user research with users with experience of domestic abuse
- user research with users with assistive tech needs
- auditing with disabled users via [DAC](https://digitalaccessibilitycentre.org/)

## Notes
This component has been built across the following pull requests:
- https://github.com/alphagov/govuk-frontend/pull/2942
- https://github.com/alphagov/govuk-frontend/pull/2943
- https://github.com/alphagov/govuk-frontend/pull/2940
- https://github.com/alphagov/govuk-frontend/pull/2949
- https://github.com/alphagov/govuk-frontend/pull/3059
- https://github.com/alphagov/govuk-frontend/pull/3081
- https://github.com/alphagov/govuk-frontend/pull/3176
- https://github.com/alphagov/govuk-frontend/pull/3209
- https://github.com/alphagov/govuk-frontend/pull/3210
- https://github.com/alphagov/govuk-frontend/pull/3247
- https://github.com/alphagov/govuk-frontend/pull/3257
- https://github.com/alphagov/govuk-frontend/pull/3221
- https://github.com/alphagov/govuk-frontend/pull/3353
- https://github.com/alphagov/govuk-frontend/pull/3305

### CODE REVIEW

File: src/govuk/components/hide-this-page/README.md
Comment (vanitabarrett) [inline src/govuk/components/hide-this-page/README.md:1]: Sorry to be picky as I know this isn't ready for production yet, but I've noticed in the past that it's really easy to forget to update this file to talk about the correct component (it's normally copied + pasted) and so sometimes we notice it a way down the line. So just an early heads up that this file still references the inset text component.

File: src/govuk/components/hide-this-page/_index.scss
Comment (vanitabarrett) [inline src/govuk/components/hide-this-page/_index.scss:26]: Nice work on this and thinking about print styles 👏

File: src/govuk/components/hide-this-page/hide-this-page.yaml
Comment (vanitabarrett) [inline src/govuk/components/hide-this-page/hide-this-page.yaml:37]: I'd suggest the default  example should only contain any required fields (so in this case, maybe nothing?)

File: src/govuk/all.js
Comment (owenatgov) [reply src/govuk/all.js:60]: In the current implementation we're planning on trying out, there would technically be 2 on a page. See the draft branch on the design system website: https://github.com/alphagov/govuk-design-system/tree/hide-this-page-component I agree that if we change it to be only one, we would only look for and initialise that 1 button.

File: src/govuk/components/hide-this-page/hide-this-page.yaml
Comment (owenatgov) [reply src/govuk/components/hide-this-page/hide-this-page.yaml:37]: I need to work out how best to handle this. Right now if I pass nothing, literally nothing gets applied ie: you get a button with not text in. We should probably set default values for this in the component itself but I've seen different implementations across the repo.

Comment (vanitabarrett) [reply src/govuk/components/hide-this-page/hide-this-page.yaml:37]: Agreed, I'm not sure if I've seen one implementation we always use for this. Potentially it depends on how complex the logic is. I think I've seen some where we do `if params.text else 'default' }}`, but there are instances where we pull out the logic to the top of the template, e.g:

```
{%- if params.text %}
    {% set buttonText = params.text %}
{%- else %}
    {% set buttonText = "default" %}
{%- endif -%}
```

Potentially there's an argument for setting defaults at the top to avoid cluttering the actual HTML code with that logic and hopefully making it easier to read.

File: src/govuk/components/hide-this-page/hide-this-page.js
Comment (querkmachine) [inline src/govuk/components/hide-this-page/hide-this-page.js:38]: [KeyboardEvent.key isn't properly supported in Internet Explorer.](https://caniuse.com/keyboardevent-key) and will need supplementing with KeyboardEvent.keyCode and/or KeyboardEvent.which. 

Something like `e.key === 'Escape' || e.keyCode === 27 || e.which === 27` outta work, I think?

File: src/govuk/components/hide-this-page/_index.scss
Comment (querkmachine) [inline src/govuk/components/hide-this-page/_index.scss:13]: Does this need to be progressively enhanced for browsers without support for `position: sticky`?

File: src/govuk/components/hide-this-page/hide-this-page.js
Comment (querkmachine) [inline src/govuk/components/hide-this-page/hide-this-page.js:37]: There are potentially some edge cases where we don't want this escape key functionality to fire, in particular, with popover elements, such as those that appear when certain form elements are in use—datepickers, select lists and the like. These could lead to users inadvertently exiting a form and losing data or progress. 

I did a quick whip around and, luckily, browser vendors seem to mostly handle that for us. (✅ = event did not propagate to `window`; ❌ = event _did_ propagate to `window`). 

|Browser|`input[type="date"]` datepicker|`select` menu|
|:-|:-|:-|
|Chrome 99|✅|✅|
|Edge 99|✅|✅|
|Opera 84|✅|✅|
|Firefox 98|✅|✅|
|Safari 15.4 (macOS)|❌|✅|
|Safari 15.4 (iOS/iPadOS)|✅|✅|
|IE 11|N/A|✅|

It's also notable that both the accessible-autocomplete and the country-and-territory-autocomplete packages currently use the Escape key to close an open popover menu, and neither prevent propagation of that event to the `window` at this time. 

Given we don't encourage the use of `input[type="date"]` anyway, the Safari issue is perhaps not the highest priority at this time. I don't know if we want to simply document these situations or attempt to resolve these in code.

File: src/govuk/components/hide-this-page/_index.scss
Comment (owenatgov) [reply src/govuk/components/hide-this-page/_index.scss:13]: It already is! (sort of). Currently on browser not supporting `sticky` it falls back to `fixed` found on line 5. There's a question about the tradeoff of having a button that follows the user but is outside the bounds of the width container and a button that doesn't follow the user but is easier to notice and access visually when it is on the screen. I'll revisit this with a designer once this has gone through some research.

File: src/govuk/components/hide-this-page/hide-this-page.js
Comment (owenatgov) [reply src/govuk/components/hide-this-page/hide-this-page.js:37]: This is incredibly thorough! Many thanks, I'll link this to the [keyboard shortcut issue](https://github.com/alphagov/govuk-design-system/issues/2116).

We're very aware of the risk of using `esc` and how it may conflict with a number of existing functionalities, however we're implementing it for now for the purpose of having an established convention for this pattern so we an test if keyboard shortcuts are actually useful or not. See [my comment on the keyboard issue](https://github.com/alphagov/govuk-design-system/issues/2116#issuecomment-1077837908) for further details.

**Comment by querkmachine** [comment]
(Realise this is an old prototype, but wanted to get this down for future reference.) 

It appears that the functionality tied to pressing `Esc` three times is blocked by Firefox's default popup blocking functionality, at least as of Firefox 105. Instead a banner is shown across the top of the viewport, indicating that the action has been blocked by the browser. No new tab is opened, nor is the current tab directed away from.

This seems to be tied to the new tab functionality. The same issue is seemingly not present on [The Trevor Project website](https://www.thetrevorproject.org/), which uses the same method. The banner warning that a new tab was blocked from opening still appears, however the current tab is then successfully navigated away from. Chromium browsers open the new tab without complaint.

Ensuring that any escape functionality works as expected with default browser settings is probably something we need to keep in mind.

**Comment by owenatgov** [comment]
@querkmachine I might have some context on this. Something we found when we were building this is that browser popup functionality to block what it asserts is malicious code includes blocking any attempts to at least open new views from a page. What turns this off is the user interacting with the page in some way ie: tapping, clicking, scrolling etc. This seems to not include keyboard interaction, meaning that if a user hasn't interacted with a page but they try to use the esc key it'll popup block. I can't tell if this is a bug or not and didn't investigate this during the prototype building.

I'm coming into this making some assumptions about what I encountered when I was last looking at this so this may not be relevant. I agree that let's keep an eye on this and see if it's still an issue when we remove the more complex tab functionality.

File: src/govuk/components/hide-this-page/hide-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/hide-this-page/hide-this-page.mjs:62]: Nice work with all this 😊 

Just a tiny tiny tiny implementation detail

Default phase for listeners is to grab events "bubbling up" (like we have here)
https://www.w3.org/TR/DOM-Level-3-Events/#event-flow

1. Capturing phase – the event goes down to the element.
2. Target phase – the event reached the target element.
3. Bubbling phase – the event bubbles up from the element.

But we're listening at the top on `window`

The `(e)` event might have already "bubbled up" through hundreds of DOM elements—potentially being cancelled/stopped by other listeners via [`e.preventDefault()`](https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault) or [`e.stopPropagation()`](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation) along the way

I'd recommend changing to the `useCapture` capturing phase (still on `window`) to ensure we grab the event on its way down, before other listeners get chance to cancel it once it bubbles back up

Comment (colinrotherham) [reply src/govuk/components/hide-this-page/hide-this-page.mjs:62]: The [GOV.UK Publishing Components **Modal dialogue**](https://github.com/alphagov/govuk_publishing_components/blob/main/app/assets/javascripts/govuk_publishing_components/components/modal-dialogue.js#L57) (good example) also listens for the escape key, but you can see they add the `useCapture` flag

They're good citizens and don't stop the event, but other listeners might do

Comment (owenatgov) [reply src/govuk/components/hide-this-page/hide-this-page.mjs:62]: Aha I missed this, why did I set this to listen on window 🤔

File: src/govuk/components/exit-this-page/exit-this-page.test.js
Comment (36degrees) [inline src/govuk/components/exit-this-page/exit-this-page.test.js:24]: Suggestion: Puppeteer has a handy [`$eval` function](https://pptr.dev/api/puppeteer.page._eval) that could be used to simplify this (and similar statements throughout these tests) slightly:

```suggestion
      const href = await page.$eval(buttonClass, el => el.href)
```

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (36degrees) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:18]: Suggestion – would it make sense for these translation key to be more aligned with their English defaults?

```suggestion
  loading: 'Loading.',
  expired: 'Exit this page expired.',
```

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (36degrees) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:24]: Can we mark everything in this class other than the constructor and init function as private? In all other classes we've deprecated everything and plan to make them private in v5 – we should  avoid these ever being 'public'.

File: app/src/views/full-page-examples/child-maintenance/index.njk
Comment (colinrotherham) [inline app/src/views/full-page-examples/child-maintenance/index.njk:59]: Tiny little stray space slipped in

```suggestion
        html: 'Births, deaths, marriages and care',
```

File: app/src/views/full-page-examples/child-maintenance/index.njk
Comment (colinrotherham) [inline app/src/views/full-page-examples/child-maintenance/index.njk:51]: We don't lint Nunjucks, but should we stick to double quotes everywhere still?

(Can use single quotes when they enclose double quotes)

```suggestion
    classes: "",
```

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (36degrees) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:130]: Suggestion: This could potentially be done in a single pass by using [`DOMTokenList.toggle`](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle):

```suggestion
  // Turn on only the indicators we want on
  var $indicators = this.$indicatorContainer.querySelectorAll(
    '.govuk-exit-this-page__indicator-light'
  )

  nodeListForEach($indicators, function ($indicator, index) {
    $indicator.classList.toggle(
      'govuk-exit-this-page__indicator-light--on',
      index < this.keypressCounter
    )
  }.bind(this))
```

File: app/src/views/full-page-examples/child-maintenance/index.njk
Comment (colinrotherham) [inline app/src/views/full-page-examples/child-maintenance/index.njk:36]: Did we want both "exit" URLs to be different? (Button uses Google versus skip link using BBC)

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (36degrees) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:147]: I don't understand what this is doing – can you explain?

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (36degrees) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:161]: Not 100% sure but might be able to avoid an extra paint by doing these the other way around?If it needs to be this way round so that it's announced it might be worth adding a comment.

File: src/govuk/all.mjs
Comment (colinrotherham) [inline src/govuk/all.mjs:63]: Shall we ditch the length check? For the other components we go straight to `nodeListForEach()`

```suggestion
```

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (36degrees) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:238]: Suggestion: Think you can avoid the extra wrapper function here:

```suggestion
    // Set a fresh timeout
  this.keypressTimeoutId = setTimeout(
    this.resetKeypressTimer.bind(this),
    this.timeoutTime
  )
```

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:33]:
```suggestion
  /** @type {ExitThisPageConfig} */
  var defaultConfig = {
```

File: src/govuk/components/exit-this-page/template.test.js
Comment (36degrees) [inline src/govuk/components/exit-this-page/template.test.js:43]: What does this mean?

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:37]: Mind adding all the JSDoc types like we do in the other components?

You'll start to see all the JavaScript type linting come to life and reveal problems if they exist

```suggestion  
  /**
   * @private
   * @type {ExitThisPageConfig}
   */
  this.config = mergeConfigs(
```

**Review by colinrotherham — CHANGES_REQUESTED** [review]

File: app/src/views/examples/exit-this-page-with-skiplink/index.njk
Comment (36degrees) [inline app/src/views/examples/exit-this-page-with-skiplink/index.njk:1]: Thanks for adding this and the full page example! Would also be good to add one to the 'Translated' example that shows the component in its translated state.

File: src/govuk/objects/_template.scss
Comment (36degrees) [inline src/govuk/objects/_template.scss:15]: I don't follow – if this is to prevent a potential issue with exit this page, why are we only adding the padding when exit this page is _not_ present?

**Review by 36degrees — COMMENTED** [review]
There's a couple of things I don't quite understand, but otherwise this is looking really good. Most of my comments are just suggestions on ways that the code could be simplified slightly – I'll leave it up to you to decide if they're worth doing.

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:18]: Good idea. This, as with a few things in the ETP code, are likely hangovers for when the content we included was something else.

File: src/govuk/components/exit-this-page/template.test.js
Comment (owenatgov) [reply src/govuk/components/exit-this-page/template.test.js]: Stale wording from when we did fancy tab stuff on click. Have since updated it.

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:130]: [Toggle has fairly meagre browser support](https://caniuse.com/?search=toggle). Are we ok putting in a polyfill?

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:147]: This is to handle the different ways that the button can be activated. If the button was clicked, we `preventDefault` and do some extra stuff before navigating. If it was activated via the keypress, we don't want to `preventDefault` as the keypress activation doesn't pass a `MouseEvent` to the the function.

We probably don't need to separate the ways we define `redirectUrl` here but we do need to have the `preventDefault` check. I'll add a clarifying comment.

File: src/govuk/objects/_template.scss
Comment (owenatgov) [reply src/govuk/objects/_template.scss:15]: I think I got the wording confused here. It should be the other way around, so don't have `scroll-padding` if ETP isn't present. It's a complicated implementation because of how we're trying to cast as wide a net as possible and mitigate as much hassle for folks as possible. I'll update the comment and extend it a bit.

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (querkmachine) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:161]: I think I flagged this at an earlier point and it was the case that the element and text need to be inserted separately for screen readers to reliably pick it up.

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:18]: On a second look I'm going to object to these changes as they most accurately represent what the text is for. "Loading", IMO, wouldn't communicate to folks using the translation API that this is for when the user activates the button. Expired I could be swayed on. Unless you have strong opinions I'm gonna skip over this one.

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:195]: I'll add a comment with Beep's point

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (36degrees) [reply src/govuk/components/exit-this-page/exit-this-page.mjs]: Gotcha, so `exitPage` can be called either:

- directly as an event handler for the `click` event
- indirectly from the event handler for the `keypress` event (`handleKeypress`)

Have you considered adding a `handleClick` function as an intermediary?

```js
ExitThisPage.prototype.handleClick = function (event) {
    event.preventDefault()
    this.exitPage()
}
```

That'd make the purpose of the `exitPage` method clearer and mean that you never need to pass it the `event` (which at the minute is optional).

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs]: So sadly it looks like we can't do this. A note from @colinrotherham on slack:

>We mainly can’t use `@private` in v4 as the `this` keyword (inside a constructor function or methods) can be `window` until it’s been instantiated. The TypeScript compiler isn’t clever enough at this sort of thing until ES2015 classes come along

The chat's continuing but we might not have a suitable workaround at the moment.

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:163]: Doh! My fault for not checking my work 🤦🏻 I'll do this then!

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:203]: Another super tiny thing

Did we want to use ~`e`~ `event` like the other components?

```suggestion
ExitThisPage.prototype.handleClick = function (event) {
```

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:171]: Just double checking, was `redirectUrl` supposed to be an optional override?

It's not used at the moment

```suggestion
ExitThisPage.prototype.exitPage = function () {
```

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:52]: We redirect to the button's `href` a bit later on:

```js
window.location.href = this.$button.getAttribute('href')
```

Think we should check `instanceof HTMLAnchorElement` then can safely access `.href` instead

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:242]: For the timeout `this.timeoutMessageId` we did a "not null" check first

Do we need one here?

Comment (colinrotherham) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:242]: Although we clear timers and intervals without checking in a few places, so maybe we're being [too strict above](https://github.com/alphagov/govuk-frontend/pull/2545/files#diff-73815616f3d4872193e2c150dfa54c35ee4232882d6b0441f749c833636a3d56R234)

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:171]: Ah whoops. This was a carryover from when I was trying out splitting how the URL we're passing to the navigation was defined. I'll get rid of it.

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:52]: I wonder if I should check this earlier on in the constructor function? We currently don't do any `$module` verification like we do for other components. I could check `$module` and `$button` at least at the outset?

Comment (colinrotherham) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:52]: Sounds brill. Yeah try and check them both before adding to `this.$xxx`

Then the compiler knows exactly what they are later 👍

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:61]: Liking the new checks ✅

Just need to swap in `$button` here

```suggestion
  this.$button = $button
```

File: src/govuk/components/exit-this-page/exit-this-page.mjs
Comment (colinrotherham) [inline src/govuk/components/exit-this-page/exit-this-page.mjs:6]: Can you just double check if we need this?

Other components that use `addEventListener()` import it

If a service team JUST imports `exit-this-page.mjs`, do they have everything they need polyfilled?

```suggestion
import '../../vendor/polyfills/Element/prototype/classList.mjs'
import '../../vendor/polyfills/Event.mjs' // addEventListener, event.target normalization and DOMContentLoaded
```

Comment (owenatgov) [reply src/govuk/components/exit-this-page/exit-this-page.mjs:6]: Let's import it just in case. Great spot.

File: src/govuk/components/exit-this-page/template.njk
Comment (owenatgov) [reply src/govuk/components/exit-this-page/template.njk:5]: I've added some extra whitespace management. I think the whitespace management is already pretty ok though. Lemme know what you think.

Comment (colinrotherham) [reply src/govuk/components/exit-this-page/template.njk:5]: I think the new `-%}` were already cared for by the next line’s `{%-` but no harm

We haven’t handled the new line before `class` or `data-module` so it’s the same as it was, sorry 😬 

That alright?

**Pull request merged**
